### ISSUE TITLE
Cannot generate regex keys when using a `Metadata` object

### ISSUE THREAD

**Issue by patmania** [sdv-dev/SDV#1160]
* SDV version: 0.17.2
* Python version: 3.9
* Operating System: Windows 11

### Problem description

I have recently started using this project, but cannot get a good grip on the primary key handling. I'd like to sample relational data using the HMA1 model. That works when i setup a metadata class and add relations and so forth. I can get samples including related (foreign key) data. Really nice.

However as soon as i say that a field is a primary key, the sampler starts creating its own primary keys based on the data type (in stead of re-sampling from what is was thought). For example my source data contains hashkeys from a datamart table, but the sampled primary keys become "a", "b", "c" ... , "Aa", ...  and so on. However in most -if not all DB systems- (and also Power BI for example) "aE" equals "AE" and "Ae" for example. So they are not really unique anymore.

And i am stuck into getting this behavior changed. I'd like to either:

1) Define primary keys (for the correct relational mappings) but keep the source keys in tact.
2) Find a way to define the format of the samples keys.

### What I already tried

There is documentation for point 2 using a regular expression, but it seems to be denied. I followed these two guides: 
https://sdv.dev/SDV/developer_guides/sdv/metadata.html
https://sdv.dev/SDV/user_guides/relational/relational_metadata.html

Below some source rows, my python code, and the samples output:

Souce keys (the name scrambling works fine, i didn't include true names in this sample source):
```
[Partner HK]	                                                                [Name]
00F59C3C92A51BB9783F6897890D22C0BC1451BA2447D72BB683BABCFA933EF9		Nathan
0231E2C076F5B397DCCF24A04707F0C0D40FF3EBAE1EBC8B71298B8CF0BC8DFF		Ronnie
0231E2C076F5B397DCCF24A04707F0C0D40FF3EBAE1EBC8B71298B8CF0BC8DFF		Patrick
0231E2C076F5B397DCCF24A04707F0C0D40FF3EBAE1EBC8B71298B8CF0BC8DFF		Curtis
0509E0556566E9A8EAA9FB020D51B4CA8CDE8304735C0FEAD55F8F6386E43C56		Dawn
0509E0556566E9A8EAA9FB020D51B4CA8CDE8304735C0FEAD55F8F6386E43C56		Dale
0509E0556566E9A8EAA9FB020D51B4CA8CDE8304735C0FEAD55F8F6386E43C56		Leah
0588C9D925DF5D1B7479883EBC52693C96EFBB6B620A64E0C53C4FD1B6CE36CB		Cheryl
076FA30A1946159243AA2579FD4BF0702D31418A7063AD3459F76335CF2FC0C4		Jennifer
```

My SDV setup (with only the minimum setup required to reproduce my error). Please note that the replacing of 'ppi' data does work, but the PK regex doesn't:

```tables = {}

tables['Partner DIM'] = pd.read_sql_table('Partner DIM', engine, 'dm')

metadata = Metadata()

metadata.add_table(
    name='Partner DIM',
    data=tables['Partner DIM'],
    primary_key='Partner HK',
    fields_metadata = {
        'Partner HK': {
            'type': 'id',
            'subtype': 'string',
            'regex': '[a-z]{24}'
        }
        , 'Name': {
            'type': 'categorical',
            'pii': True,
            'pii_category': 'first_name'
        }
    }
)

model = HMA1(metadata)
model.fit(tables)
data = model.sample()
```

And finally some output:

```[Partner HK]	[Name]
a			Jeffrey
A			Maria
aA			Javier
aa			Jesse
ab			Robin
aB			Jesse
aC			Benjamin
ac			David
ad			Michelle
```

Which already shows 'aA' and 'aa' for example.

Thanks a lot for any insights!

**Comment by patmania** [comment]
I have a bit more details. I am now debugging and it seems that the part in the attachment in file dataset.py is deleting my metadata by overwriting my metadata with the field data. After this part, my 'regex' field is gone. I am trying to confirm this with a unit test now. There are unit tests for the regex, but they only test the 'fields' part of the meta data, not an entire table setup. I will keep you posted..
![2023_01_06_18_06_37_dataset py_DataSynth_Visual_Studio_Code](https://user-images.githubusercontent.com/122096391/211062127-217291b9-af69-4af0-b1de-d626eb13b8a6.png)

**Comment by patmania** [comment]
I can confirm that it has to do with the Metadata.add_table method. If i pass a json dictionary as the metadata argument (such as these examples: https://sdv.dev/SDV/developer_guides/sdv/metadata.html) to my modeler, it does work.

I will continue my investigation and update this ticket. I also hope that i can fix it.

**Comment by npatki** [comment]
Hi @patmania thanks for filing this issue! For now, I'd suggest the workaround of storing & passing in the metadata separately as a `json` object rather than using the `Metadata` class to construct it dynamically.

Note that an upcoming release of SDV will come with significant improvements to the `Metadata` generation, so I suspect you'll be able to resume your intended workflow then.

For now, I'll mark this as a bug and we can keep it open to track until the new release. I'll also update the title to reflect the new details.

**Comment by patmania** [comment]
Hi @npatki thanks for reaching out! I have moved on by using a workaround for now. After setting up my metadata instance with 'add_table' and 'add_relation' calls i convert it to a dictionary and add my regex expressions to it:

```
meta_dict = metadata.to_dict()
meta_dict['tables']['FACT']['fields']['Fact Key'] = { 'type': 'id', 'subtype': 'string', 'regex': '[a-z]{10}' }

model = HMA1(meta_dict)
```

So i still have the convenience of add_table by not having to specify all my dozens of fields manually :)

**Comment by npatki** [comment]
Awesome! If it helps, many users also save the metadata dictionary separately as a JSON file. We'll be recommending this approach in the future.

```python
import json

# write to file
with open('metadata.json', 'w') as file:
  json.dump(metadata_dict, file)

# read from file
with open('metadata.json') as file:
  metadata_dict = json.load(file)
```

**Comment by npatki** [comment]
Great news! This issue has now been resolved in the new, [**SDV 1.0 (Beta!)**](https://docs.sdv.dev/sdv) release.

Note that SDV 1.0 has an updated format of metadata as well as a new set of tools to help you create it. For more information see our [documentation site](https://docs.sdv.dev/sdv/multi-table-data/data-preparation),

You may also read through our tutorial notebooks for [creating metadata](https://colab.research.google.com/drive/1zGqvHYcfhkBmVVxvPhVV0Htq4XkyhKkB?usp=sharing) and [synthesizing multiple tables](https://colab.research.google.com/drive/1h_xwzHXrif-xEOC39amYC6xfNPu5hDA1?usp=sharing).
